builtins: introduce with_min_timestamp and with_max_staleness

otan opened the pull request:
Introduce the `with_min_timestamp` and `with_max_staleness` builtins,
both returning a TimestampTZ of the minimum bound. The RFC has been
updated accordingly.

These functions are gated by a CCL enterprice license, and the relevant
dependency injections have been added to account for that.

The "fail fast" method discussed I plan to add in a later PR.

Refs: #67555

Release note (sql change): Introduce `with_min_timestamp` and
`with_max_staleness` builtins. In a SELECT clause, they return the same
timestamp and (now() - interval), but are intended for use in AS OF
SYSTEM TIME which will appear in an upcoming update.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67697)
<!-- Reviewable:end -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 13 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67697#-:-Mel3yy_Eny8mXPZ7dbp:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/ccl/kvccl/kvboundedstalenessccl/kvboundedstalenessccl.go, line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67697#-Mel2YdXDsp_-Quz10Jv:-Mel2YdXDsp_-Quz10Jw:bw9065i) ([raw file](https://github.com/cockroachdb/cockroach/blob/db10c74bbe72fbe209ab4edfb42303adf897efaf/pkg/ccl/kvccl/kvboundedstalenessccl/kvboundedstalenessccl.go#L9)):*
> ```Go
> //     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
> 
> package kvboundedstalenessccl
> ```

Consider putting all of this in `kvfollowerreadsccl`. Bounded staleness reads are a form of follower reads, and if anything, that package is already a bit narrow.

If you want, you could pull out a new `builtins.go` file and move `evalFollowerReadOffset` over to it. You may even want to make the CCL injection of `EvalFollowerReadOffset` and these two builtins a little more uniform one way or another.
___
*[pkg/ccl/kvccl/kvboundedstalenessccl/kvboundedstalenessccl.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67697#-Mel3FyuD2RbdPvUtPa9:-Mel3FyuD2RbdPvUtPaA:b-7i8dyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/db10c74bbe72fbe209ab4edfb42303adf897efaf/pkg/ccl/kvccl/kvboundedstalenessccl/kvboundedstalenessccl.go#L44)):*
> ```Go
> 		)
> 	}
> 	return duration.Add(ctx.GetTxnTimestamp(time.Microsecond).Time, d.Mul(-1)), nil
> ```

Is `TxnTimestamp` what we want? We use `StmtTimestamp` in `EvalFollowerReadOffset`.  Is that one wrong?
___
*[pkg/ccl/kvccl/kvboundedstalenessccl/kvboundedstalenessccl_test.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67697#-Mel22WrEm9qNEtQAhNU:-Mel22WrEm9qNEtQAhNV:b-cvfdga) ([raw file](https://github.com/cockroachdb/cockroach/blob/db10c74bbe72fbe209ab4edfb42303adf897efaf/pkg/ccl/kvccl/kvboundedstalenessccl/kvboundedstalenessccl_test.go#L39)):*
> ```Go
> 		t.Run(q, func(t *testing.T) {
> 			_, err := tc.Conns[0].QueryContext(ctx, q)
> 			require.Error(t, err)
> ```

Is it worth sanity checking that these functions do not return this error when the license is enabled. I know you're already doing that in the logic test, but we might as well make this test self-contained for that one condition.
___
*[pkg/ccl/logictestccl/testdata/logic_test/bounded_staleness, line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67697#-Mel1v768a8_TQ6Kke4I:-Mel1v768a8_TQ6Kke4J:b-6h7iic) ([raw file](https://github.com/cockroachdb/cockroach/blob/db10c74bbe72fbe209ab4edfb42303adf897efaf/pkg/ccl/logictestccl/testdata/logic_test/bounded_staleness#L2)):*
> ```
> query B
> SELECT with_min_timestamp(now()) = now()
> ```

Clever way to make this deterministic.
___
*[pkg/sql/logictest/testdata/logic_test/bounded_staleness, line 10 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67697#-Mel1gvj2_eUQFfWhzSu:-Mel1gvj2_eUQFfWhzSv:b1ll9ku) ([raw file](https://github.com/cockroachdb/cockroach/blob/db10c74bbe72fbe209ab4edfb42303adf897efaf/pkg/sql/logictest/testdata/logic_test/bounded_staleness#L10)):*
> ```
> 
> statement error pgcode XXC01 with_max_staleness can only be used with a CCL distribution
> SELECT with_max_staleness('1s'::interval)
> ```

nit: do we need the cast?
___
*[pkg/sql/sem/builtins/builtins.go, line 7620 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67697#-Mel0WSq5wOaVmFWmk0G:-Mel0WSq5wOaVmFWmk0H:baqm9u3) ([raw file](https://github.com/cockroachdb/cockroach/blob/db10c74bbe72fbe209ab4edfb42303adf897efaf/pkg/sql/sem/builtins/builtins.go#L7620)):*
> ```
> an single-statement
> ```

"a single-statement" here and below.
___
*[pkg/sql/sem/builtins/builtins.go, line 7621 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67697#-Mel0_xcDqa7fj_wtrBI:-Mel0_xcDqa7fj_wtrBJ:b-6ftjyl) ([raw file](https://github.com/cockroachdb/cockroach/blob/db10c74bbe72fbe209ab4edfb42303adf897efaf/pkg/sql/sem/builtins/builtins.go#L7621)):*
> ```
> before the min_timestamp
> ```

equal to or later than the min_timestamp


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67697)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[pkg/ccl/kvccl/kvboundedstalenessccl/kvboundedstalenessccl.go, line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67697#-Mel2YdXDsp_-Quz10Jv:-MelFGBD3mYQYerRdUgt:b-ps0jr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/db10c74bbe72fbe209ab4edfb42303adf897efaf/pkg/ccl/kvccl/kvboundedstalenessccl/kvboundedstalenessccl.go#L9)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider putting all of this in `kvfollowerreadsccl`. Bounded staleness reads are a form of follower reads, and if anything, that package is already a bit narrow.

If you want, you could pull out a new `builtins.go` file and move `evalFollowerReadOffset` over to it. You may even want to make the CCL injection of `EvalFollowerReadOffset` and these two builtins a little more uniform one way or another.
</blockquote></details>

I'll move things over there for now, the EvalFollowerReadOffset seems a little bit more complicated because it also has to deal with CCL and enterprise not being enabled.
___
*[pkg/ccl/kvccl/kvboundedstalenessccl/kvboundedstalenessccl.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67697#-Mel3FyuD2RbdPvUtPa9:-Mev_BVL2DqpJweKX3wH:b-6r7id7) ([raw file](https://github.com/cockroachdb/cockroach/blob/db10c74bbe72fbe209ab4edfb42303adf897efaf/pkg/ccl/kvccl/kvboundedstalenessccl/kvboundedstalenessccl.go#L44)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is `TxnTimestamp` what we want? We use `StmtTimestamp` in `EvalFollowerReadOffset`.  Is that one wrong?
</blockquote></details>

as discussed on slack, we'll use StmtTimestamp. I've made this a separate commit to make it easy to revert :P


<!-- Sent from Reviewable.io -->

otan commented:
thanksss

bors r=nvanbenschoten

otan commented:
bors r=nvanbenschoten

otan committed:
builtins: introduce with_min_timestamp and with_max_staleness

Introduce the `with_min_timestamp` and `with_max_staleness` builtins,
both returning a TimestampTZ of the minimum bound. The RFC has been
updated accordingly.

These functions are gated by a CCL enterprice license, and the relevant
dependency injections have been added to account for that.

The "fail fast" method discussed I plan to add in a later PR.

Release note (sql change): Introduce `with_min_timestamp` and
`with_max_staleness` builtins. In a SELECT clause, they return the same
timestamp and (now() - interval), but are intended for use in AS OF
SYSTEM TIME which will appear in an upcoming update.

otan committed:
kvfollowerreadsccl: switch to statement_timestamp

Switch from using TxnTimestamp to StmtTimestamp. Separate commit so it
is easily revertible.

Release note: None

otan commented:
oooft, ok that was an obscure CI only bug (need to round StatementTimestamp in code!)

bors r=nvanbenschoten
